aten/src/ATen/native/Activation.cpp

```diff
@@ -1,33 +1,267 @@
+        output_data[d * dim_stride] = z;
+      }
+    }
+
+    if (LogSoftMax)
+      tmpsum = max_input + std::log(tmpsum);
+    else
+      tmpsum = 1 / tmpsum;
+
+    for (d = 0; d <  dim_size; d++)
+      if (LogSoftMax)
+        output_data[d * dim_stride] = input_data[d * dim_stride] - tmpsum;
+      else
+        output_data[d * dim_stride] *= tmpsum;
+  }
+}
+
+template <typename scalar_t, bool LogSoftMax>
+void host_softmax_backward(
+    Tensor& gI,
+    const Tensor& grad,
+    const Tensor& output,
+    int64_t dim) {
+
+  int64_t outer_size = 1;
+  int64_t dim_size = grad.size(dim);
+  int64_t inner_size = 1;
+  for (int64_t i = 0; i < dim; ++i)
+    outer_size *= grad.size(i);
+  for (int64_t i = dim + 1; i < grad.dim(); ++i)
+    inner_size *= grad.size(i);
+  int64_t dim_stride = inner_size;
+  int64_t outer_stride = dim_size * dim_stride;
+  scalar_t* gradInput_data_base = gI.data<scalar_t>();
+  scalar_t* output_data_base = output.data<scalar_t>();
+  scalar_t* gradOutput_data_base = grad.data<scalar_t>();
+  int64_t i, d;
+
+#pragma omp parallel for private(i, d)
+  for (i = 0; i < (outer_size * inner_size); i++) {
+    int64_t outer_idx = i / inner_size;
+    int64_t inner_idx = i % inner_size;
+    scalar_t* gradInput_data =
+        gradInput_data_base + outer_idx * outer_stride + inner_idx;
+    scalar_t* output_data =
+        output_data_base + outer_idx * outer_stride + inner_idx;
+    const scalar_t* gradOutput_data =
+        gradOutput_data_base + outer_idx * outer_stride + inner_idx;
+
+    scalar_t sum = 0; // TODO was accreal here
+  for (int ii = 0; ii < dim_size; ii++) {
+  }
+    for (d = 0; d <  dim_size; d++)
+      if (LogSoftMax)
+        sum += gradOutput_data[d * dim_stride];
+      else
+        sum += gradOutput_data[d * dim_stride] * output_data[d * dim_stride];
+
+
+    for (d = 0; d < dim_size; d++) {
+      if (LogSoftMax) {
+        gradInput_data[d * dim_stride] = gradOutput_data[d * dim_stride] -
+            std::exp(output_data[d * dim_stride]) * sum;
+      } else {
+        gradInput_data[d * dim_stride] = output_data[d * dim_stride] *
+            (gradOutput_data[d * dim_stride] - sum);
+      }
+    }
+  }
+}
+} // namespace
+
+Tensor softmax_cpu(const Tensor& input_, const int64_t dim_) {
+  auto input = input_.contiguous();
+  Tensor output = at::native::empty_like(input);
+  Tensor output2 = at::native::empty_like(input);
+  int64_t dim = maybe_wrap_dim(dim_, input.dim());
+  if (input.dim() == 0)
+    input = input.view(1);
+  AT_CHECK(
+      dim >= 0 && dim < input.dim(),
+      "dim must be non-negative and less than input dimensions");
+  if (input.ndimension()  > 0 && dim == input.ndimension() - 1) {
+    softmax_lastdim_kernel(output, input);
+  } else {
+    AT_DISPATCH_FLOATING_TYPES(input.type(), "softmax", [&] {
+      host_softmax<scalar_t, false>(output, input, dim);
+    });
+  }
+  return output;
+}
+
+Tensor log_softmax_cpu(const Tensor& input_, const int64_t dim_) {
```

Comment:
Can bodies of the forward/backward functions be templated too? They look pretty much the same to me.

Reply:
Sure, it's only tricky for log_softmax since that needs to use another layer of looping to make use of the vectorization of .log. We can't call std::log from avx2 code because of the glibc bug.